lib/ipam/ipam.go

```diff
@@ -785,26 +784,36 @@ func (c ipamClient) AssignIP(ctx context.Context, args AssignIPArgs) error {
 		}
 
 		// Increment handle.
+		var h *model.KVPair
 		if args.HandleID != nil {
-			c.incrementHandle(ctx, *args.HandleID, blockCIDR, 1, args.Attrs)
+			h, err = c.incrementHandle(ctx, *args.HandleID, blockCIDR, 1, args.Attrs)
+			if err != nil {
+				if _, ok := err.(cerrors.ErrorResourceUpdateConflict); ok {
+					log.WithError(err).Debug("CAS error incrementing handle - retry")
+					continue
+				}
+				log.WithError(err).Errorf("Failed to increment handle")
+				return err
+			}
 		}
 
 		// Update the block using the original KVPair to do a CAS.  No need to
 		// update the Value since we have been manipulating the Value pointed to
 		// in the KVPair.
 		_, err = c.blockReaderWriter.updateBlock(ctx, obj)
 		if err != nil {
+			log.WithError(err).Warningf("Update failed on block %s", block.CIDR.String())
+			if h != nil {
+				// Attempt to decrement the handle that was incremented above, so the handle stays
+				// consistent with allocation state.
+				if _, decErr := c.decrementHandle(ctx, h, blockCIDR, 1); decErr != nil {
+					log.WithError(decErr).Warn("Failed to decrement handle")
+				}
+			}
 			if _, ok := err.(cerrors.ErrorResourceUpdateConflict); ok {
 				log.WithError(err).Debug("CAS error assigning IP - retry")
 				continue
 			}
-
```

Comment:
Moved this up a few lines so that we decrement the handle even if we hit a CAS on the block (since we'll increment it on retry so we need to also decrement it here).